File: macos/Classes/VolumeController.swift
//
//  VolumeController.swift
//  flutter_volume_controller
//
//  Created by yosemiteyss on 17/9/2022.
//

import Foundation
import FlutterMacOS

class VolumeController {
    func getVolume() throws -> Float {
        return try SoundOutputManager.shared.readVolume()
    }
    
    func setVolume(_ volume: Float) throws {
        try SoundOutputManager.shared.setVolume(volume)
    }
    
    func raiseVolume(_ step: Float?) throws {
        try SoundOutputManager.shared.increaseVolume(by: step ?? 0.15)
    }
    
    func lowerVolume(_ step: Float?) throws {
        try SoundOutputManager.shared.decreaseVolume(by: step ?? 0.15)
    }
    
    func getMute() throws -> Bool {
        return try SoundOutputManager.shared.readMute();
    }
    
    func setMute(_ isMuted: Bool) throws {
        try SoundOutputManager.shared.mute(isMuted)
    }
    
    func toggleMute() throws {
        let isMuted = try getMute()
        try setMute(!isMuted)
    }
    
    func getDefaultOutputDevice() throws -> OutputDevice {
        return try SoundOutputManager.shared.retrieveDefaultOutputDevice()
    }
    
    func setDefaultOutputDevice(_ deviceId: String) throws {
        try SoundOutputManager.shared.setDefaultOutputDevice(deviceId)
    }
    
    func getOutputDeviceList() throws -> [OutputDevice] {
        return try SoundOutputManager.shared.retrieveOutputDeviceList()
    }
}

File: android/src/main/kotlin/com/yosemiteyss/flutter_volume_controller/VolumeController.kt
package com.yosemiteyss.flutter_volume_controller

import android.media.AudioManager
import android.os.Build

class VolumeController(private val audioManager: AudioManager) {
    fun getVolume(audioStream: AudioStream): Double {
        return audioManager.getVolume(audioStream)
    }

    fun setVolume(volume: Double, showSystemUI: Boolean, audioStream: AudioStream) {
        val max = audioManager.getStreamMaxVolume(audioStream.streamType)
        audioManager.setStreamVolume(
            audioStream.streamType,
            (max * volume).toInt(),
            if (showSystemUI) AudioManager.FLAG_SHOW_UI else 0
        )
    }

    fun raiseVolume(step: Double?, showSystemUI: Boolean, audioStream: AudioStream) {
        if (step == null) {
            audioManager.adjustStreamVolume(
                audioStream.streamType,
                AudioManager.ADJUST_RAISE,
                if (showSystemUI) AudioManager.FLAG_SHOW_UI else 0
            )
        } else {
            val target = getVolume(audioStream) + step
            setVolume(target, showSystemUI, audioStream)
        }
    }

    fun lowerVolume(step: Double?, showSystemUI: Boolean, audioStream: AudioStream) {
        if (step == null) {
            audioManager.adjustStreamVolume(
                audioStream.streamType,
                AudioManager.ADJUST_LOWER,
                if (showSystemUI) AudioManager.FLAG_SHOW_UI else 0
            )
        } else {
            val target = getVolume(audioStream) - step
            setVolume(target, showSystemUI, audioStream)
        }
    }

    fun getMute(audioStream: AudioStream): Boolean {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            audioManager.isStreamMute(audioStream.streamType)
        } else {
            audioManager.getStreamVolume(audioStream.streamType) == 0
        }
    }

    @Suppress("DEPRECATION")
    fun setMute(isMuted: Boolean, showSystemUI: Boolean, audioStream: AudioStream) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            audioManager.adjustStreamVolume(
                audioStream.streamType,
                if (isMuted) AudioManager.ADJUST_MUTE else AudioManager.ADJUST_UNMUTE,
                if (showSystemUI) AudioManager.FLAG_SHOW_UI else 0
            )
        } else {
            audioManager.setStreamMute(audioStream.streamType, isMuted)
        }
    }

    fun toggleMute(showSystemUI: Boolean, audioStream: AudioStream) {
        val isMuted = getMute(audioStream)
        setMute(!isMuted, showSystemUI, audioStream)
    }
}


